File: source/RolXServer/RolXServer/Reports/WebApi/Resource/ReportEntry.cs
// -----------------------------------------------------------------------
// <copyright file="ReportEntry.cs" company="Christian Ewald">
// Copyright (c) Christian Ewald. All rights reserved.
// Licensed under the MIT license.
// See LICENSE.md in the project root for full license information.
// </copyright>
// -----------------------------------------------------------------------

namespace RolXServer.Reports.WebApi.Resource;

/// <summary>
/// A single entry of a report.
/// </summary>
public sealed record ReportEntry(
    string Date,
    int ProjectNumber,
    string CustomerName,
    string ProjectName,
    int SubprojectNumber,
    string SubprojectName,
    int ActivityNumber,
    string ActivityName,
    string BillabilityName,
    string UserName,
    long Duration,
    string Comment);

File: source/rolx-client/src/app/reports/core/report-entry.ts
import { Duration, TransformAsDuration } from '@app/core/util/duration';
import { TransformAsIsoDate } from '@app/core/util/iso-date';
import { assertDefined } from '@app/core/util/utils';
import * as moment from 'moment';

export class ReportEntry {
  @TransformAsIsoDate()
  date!: moment.Moment;

  projectNumber!: number;
  customerName!: string;
  projectName!: boolean;
  subprojectNumber!: number;
  subprojectName!: string;
  activityNumber!: number;
  activityName!: string;
  billabilityName!: string;
  userName!: string;

  @TransformAsDuration()
  duration!: Duration;

  comment!: string;

  validateModel(): void {
    assertDefined(this, 'date');
    assertDefined(this, 'projectNumber');
    assertDefined(this, 'customerName');
    assertDefined(this, 'projectName');
    assertDefined(this, 'subprojectNumber');
    assertDefined(this, 'subprojectName');
    assertDefined(this, 'activityNumber');
    assertDefined(this, 'activityName');
    assertDefined(this, 'billabilityName');
    assertDefined(this, 'userName');
    assertDefined(this, 'duration');
    assertDefined(this, 'comment');
  }
}
